ledger/src/blockstore.rs

```diff
@@ -5556,6 +5621,67 @@ pub mod tests {
             }
         }
     }
+
+    #[test]
+    fn test_set_and_chain_connected_on_root_and_next_slots() {
+        let ledger_path = get_tmp_ledger_path_auto_delete!();
+        let blockstore = Blockstore::open(ledger_path.path()).unwrap();
+
+        // Create enough entries to ensure 5 shreds result
+        let entries_per_slot = max_ticks_per_n_shreds(5, None);
+
+        // Start a chain from a slot not in blockstore, such is the case when
+        // a node starts with no blockstore and downloads a snapshot. This is
+        // the special case and slot will be marked connected despite its'
+        // parent not being connected (or existing) and not being full
+        blockstore
+            .set_and_chain_connected_on_root_and_next_slots(5)
+            .unwrap();
+        let slot_meta5 = blockstore.meta(5).unwrap().unwrap();
+        assert!(!slot_meta5.is_full());
+        assert!(slot_meta5.is_parent_connected());
+        assert!(slot_meta5.is_connected());
+
+        // Insert some slots and ensure they connect from slot 5 correctly
+        let (shreds, _) = make_many_slot_entries(6, 5, entries_per_slot);
+        blockstore.insert_shreds(shreds, None, false).unwrap();
+        for slot in 6..=10 {
+            let meta = blockstore.meta(slot).unwrap().unwrap();
+            assert!(meta.is_parent_connected());
+            assert!(meta.is_connected());
+        }
+
+        // Chain connected on slots that are already connected, should just noop
+        blockstore
+            .set_and_chain_connected_on_root_and_next_slots(6)
+            .unwrap();
+        for slot in 6..=10 {
+            let meta = blockstore.meta(slot).unwrap().unwrap();
+            assert!(meta.is_parent_connected());
+            assert!(meta.is_connected());
+        }
+
+        // Insert some slots that are not connected and then ensure they connect
+        // This might be the case for a node that had existing blockstore data
+        // prior to the is_connected behavior being expanded to snapshot slots
+        let (shreds, _) = make_many_slot_entries(15, 5, entries_per_slot);
+        blockstore.insert_shreds(shreds, None, false).unwrap();
+        for slot in 15..=19 {
+            let meta = blockstore.meta(slot).unwrap().unwrap();
+            assert!(!meta.is_parent_connected());
+            assert!(!meta.is_connected());
+        }
+        blockstore
+            .set_and_chain_connected_on_root_and_next_slots(15)
+            .unwrap();
+        for slot in 15..=19 {
```

Comment:
Can we also make some slots not full (e.g. 18) here and check we only have is_connected() and is_parent_connected() up to slot 17, and slot 18 has is_parent_connected() true but is_connected() false?

Reply:
And then make the slot full, and invoke the function again to make sure every slot is both connected and parent_connected?

Reply:
Good call, done